# Table of Contents
    ## Mixed-type data
    ## Missing Values
    ## Duplicates
    ## Tidying up and exporting changes

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#create path to project folder
path = r'C:\Users\mmreg\OneDrive\Desktop\Data Analytics Course Work\Data Immersion\Tasks\04-2022 Instacart Basket Analysis'

In [11]:
#import orders csv
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'IC Orders.csv'), index_col = False)

In [12]:
#import product csv
df_prod = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'IC Products.csv'), index_col = False)

# Mixed-Type Data

In [3]:
#create a dataframe
df_test = pd.DataFrame()

In [4]:
#create a mixed-type column
df_test['mixed'] = ['a', 'b', 1, True]

In [5]:
#check work
df_test.head()

mixed
0     a
1     b
2     1
3  True

In [10]:
#look for mixed-typed data
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird]) > 0:
        print(col)

mixed


# Missing Values

In [13]:
#sum total NaN values in dataframe
df_prod.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
#create subset with only missing values
df_nan = df_prod[df_prod['product_name'].isnull() == True]

In [15]:
#view new subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
#view number of rows in current dataframe
df_prod.shape

(49693, 5)

In [20]:
#create new df with NaN values removed
df_prod_clean = df_prod[df_prod['product_name'].isnull() == False]

In [21]:
df_prod_clean.shape #shows removal was successful

(49677, 5)

# Duplicates

In [22]:
#create subset with duplicate records
df_dups = df_prod_clean[df_prod_clean.duplicated()]

In [23]:
#run new duplicate subset
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
#find shape of clean product dataframe
df_prod_clean.shape

(49677, 5)

In [26]:
#create subset of clean product dataframe without duplicates, and check new shape
df_prod_clean_no_dupes = df_prod_clean.drop_duplicates()
df_prod_clean_no_dupes.shape #shape shows all five duplicates have been dropped

(49672, 5)

In [27]:
#export cleaned product dataset
df_prod_clean_no_dupes.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 4.5 Task

## Question 2

In [28]:
#gain descriptive statistics on df_prod
df_prod.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

### The first thing that comes to my attention is the max value in the prices. 99999 seems very expensive for the products that are in this dataframe. Another thing of note is that there should be 49,693 unique product id numbers (as seen in the count figure), however when we look at the max number it is showing we are missing 5 id numbers. We have cleaned the dataset in the exercise, but this shows that it could be identified by running this step first. Everything else seems to be as expected.

## Question 3

In [30]:
#identify any mixed type data in orders dataset
for col in df_orders.columns.tolist():
  weird = (df_orders[[col]].applymap(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders[weird]) > 0:
    print (col)

### There are no mixed-type values in the orders csv

## Question 4

### As there are no mixed-type values in the dataset, no modifications are required

## Question 5

In [31]:
#identify missing values in orders dataset
df_orders.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### The only missing values within the orders dataset are in the days_since_prior_order column. This is to be expected however, as each user in the dataset would have to have a first order with Instacart; this would return no days since prior order.

## Question 6

In [49]:
#create new dataset for cleaned data with new column to identify first orders
df_orders_cleaned = df_orders

In [50]:
#add new column to identify whether the order is the first or not
df_orders_cleaned['first_order'] = df_orders['days_since_prior_order'].isnull() == True

In [51]:
#check to ensure column as been added
df_orders_cleaned.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  first_order  
0                     NaN         True  
1                    15.0        False  
2                    21.0        False  
3                    29.0        False  
4                    28.0        False

### We cannot remove the information as it is pertinent to the operations of the company, and we cannot find a mean/median to impute as it has importance on its own merits. Therefore, we have to add a new variable, which identifies whether the order in question was the first order for the user_id or not.

## Question 7 and 8

In [44]:
#identify duplicate data in orders dataset
df_order_duplicates = df_orders[df_orders.duplicated()]
print(df_order_duplicates)

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []


In [52]:
#see if there are duplicates
df_order_duplicates.shape

(0, 7)

### Based on the shape of the new subset, we can see that there are no duplicates within the order dataset. Therefore, no actions are needed to rectify a duplicate situation

## Question 9

In [53]:
#export cleaned orders dataset
df_orders_cleaned.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [54]:
#export cleaned products dataset
df_prod_clean_no_dupes.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))